src/plugins/interactive_setup/server/plugin.ts

```diff
@@ -0,0 +1,155 @@
+    const isConfigured =
+      core.elasticsearch.config.credentialsSpecified && core.elasticsearch.config.hosts.length;
+
+    /**
+     * env: dev -> enabled: false -> skip
+     * env: dev -> enabled: undefined -> skip
+     * env: dev -> enabled: true -> hold setup
+     * env: prod -> enabled: false -> skip
+     * env: prod -> enabled: undefined -> hold setup if not configured
+     * env: prod -> enabled: true -> hold setup if not configured
+     */
+    const shouldHoldSetupPhase = isDev
+      ? enabled === true
+      : enabled !== false
+      ? !isConfigured
+      : false;
+
+    if (!shouldHoldSetupPhase) {
+      return;
+    }
+
+    core.http.registerRoutes('', (prebootRouter) => {
+      let completeSetup: (result: { shouldReloadConfig: boolean }) => void;
+
+      prebootRouter.post(
+        {
+          path: '/api/preboot/setup',
+          validate: {
+            body: schema.oneOf([
+              schema.object({
+                hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+                apiKey: schema.string(),
+                caFingerprint: schema.string(),
+              }),
+              schema.object({
+                hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+                username: schema.string(),
+                password: schema.string(),
+                caFingerprint: schema.string(),
+              }),
+            ]),
+          },
+          options: { authRequired: false },
+        },
+        async (context, request, response) => {
+          if (!core.preboot.isSetupOnHold()) {
+            return response.badRequest();
+          }
+
+          const client = core.elasticsearch
+            .createClient('data', {
+              hosts: request.body.hosts.map((host) => `https://${host}`),
+              ssl: { verificationMode: 'none' },
+              // caFingerprint: request.body.caFingerprint,
+            })
+            .asScoped({
+              headers: {
+                authorization:
+                  'apiKey' in request.body
+                    ? `ApiKey ${btoa(request.body.apiKey)}`
+                    : `Basic ${btoa(`${request.body.username}:${request.body.password}`)}`,
+              },
+            });
+
+          try {
+            const { body } = await client.asCurrentUser.transport.request({
+              method: 'GET',
+              path: '/_security/enroll/kibana',
+            });
+
+            const configPath = isDev
+              ? this.#initializerContext.env.configs.find((path) => path.includes('dev'))
+              : this.#initializerContext.env.configs[0];
+
+            if (!configPath) {
+              return response.customError({ statusCode: 500, body: 'Cannot find config file.' });
+            }
+
+            await fs.appendFile(
+              configPath,
+              `
+
+elasticsearch.hosts: [${request.body.hosts.map((host) => `"https://${host}"`).join(', ')}]
+elasticsearch.username: "kibana_system"
```

Comment:
__note:__ in case of manual setup `username` should be provided by the user (we may prefill username field in UI with `kibana_system` though).

Reply:
How would the user know the kibana_system user credentials? They might not even be created at that point. 

My understanding was that you'd enter elastic user credentials and we enroll Kibana the same way we do using the enrolment token.

Reply:
Nothing is set in stone here yet, but let me dump what I had in mind.

> How would the user know the kibana_system user credentials? They might not even be created at that point.

My understanding was that users would use this tool https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-passwords.html before they set up Kibana. I think that's what users do today.

> My understanding was that you'd enter elastic user credentials and we enroll Kibana the same way we do using the enrolment token.

Hmm, do we even have a HTTP API to generate Kibana enrollment tokens? I also was thinking about the case when ES isn't not configured to use HTTPs or when Enrollment is explicitly disabled...

Reply:
You don't need an enrolment token to call the enrollment endpoint. The enrolment token contains an API key with permission to enroll kibana but the elastic (superuser) also has them so we can call the endpoint using basic auth instead of an API key. 

I think this massively simplifies getting started since you don't have to worry about creating credentials for kibana system user.

Reply:
> I think this massively simplifies getting started since you don't have to worry about creating credentials for kibana system user.

There are a number of requirements that should be met though, which would not be required otherwise:

* User should have `elastic` user credentials
* Enrollment API should be enabled in Elasticsearch
* It works only if Elasticsearch is configured to use TLS

Maybe it's okay, I'm just not sure yet how I feel about that.

Reply:
Thanks for raising these. I agree with your concerns and have changed to require kibana system user now. 

However, I still have concerns about users having to manually create kibana system user credentials outside the interactive setup flow in order to configure Kibana. I'll double check with ES team when Kibana enrolment API is active. During local testing it seems to be working even if enrolment mode is disabled on ES side.